### PR TITLE
*: set missing CLOEXEC on opened fds

### PR DESCRIPTION
**Pull request by batrick**
Otherwise these descriptors may leak across execve() during e.g. MDS respawn.
    
Fixes: http://tracker.ceph.com/issues/35850

### CODE REVIEW

**Comment by batrick** [comment]
Still a few leaks to locate and seal.

**Comment by batrick** [comment]
Our use of `std::random_device`/`boost::random::random_device` also leaks a file descriptor to `/dev/urandom`. This is not easily fixed. @cbodley @adamemerson any suggestions? Here's a patch I was using for `src/include/uuid.h` but I missed that `src/include/random.h` also needs fixed.

````
diff --git a/src/include/uuid.h b/src/include/uuid.h
index 219def4b8a..6ac16d90a4 100644
--- a/src/include/uuid.h
+++ b/src/include/uuid.h
@@ -6,12 +6,16 @@
  */
 
 #include "encoding.h"
+#include "include/assert.h"
+
 #include <ostream>
+#include <random>
+
+#include <unistd.h>
 
 #include <boost/uuid/uuid.hpp>
 #include <boost/uuid/uuid_generators.hpp>
 #include <boost/uuid/uuid_io.hpp>
-#include <boost/random/random_device.hpp>
 
 struct uuid_d {
   boost::uuids::uuid uuid;
@@ -26,9 +30,7 @@ struct uuid_d {
   }
 
   void generate_random() {
-    boost::random::random_device rng("/dev/urandom");
-    boost::uuids::basic_random_generator<boost::random::random_device> gen(&rng);
-    uuid = gen();
+    uuid = random_gen();
   }
   
   bool parse(const char *s) {
@@ -44,7 +46,7 @@ struct uuid_d {
     memcpy(s, boost::uuids::to_string(uuid).c_str(), 37);
   }
 
- std::string to_string() const {
+  std::string to_string() const {
     return boost::uuids::to_string(uuid);
   }
 
@@ -59,6 +61,26 @@ struct uuid_d {
   void decode(bufferlist::const_iterator& p) const {
     ::decode_raw(uuid, p);
   }
+
+private:
+  using prng = std::mt19937_64;
+  static prng::result_type get_seed() {
+    /* Don't feel tempted to use std::random_device. It does not use O_CLOEXEC
+     * which can cause file descriptors to leak (during daemon respawn).
+     */
+    decltype(random)::result_type r;
+    int fd = open("/dev/urandom", O_RDONLY|O_CLOEXEC);
+    if (fd < 0) {
+      ceph_abort();
+    }
+    ssize_t count = read(fd, &r, sizeof(r));
+    ceph_assert(count > 0 && (size_t)count == sizeof(r));
+    close(fd);
+    return r;
+  }
+
+  inline static prng random{get_seed()};
+  inline static boost::uuids::basic_random_generator random_gen{random};
 };
 WRITE_CLASS_ENCODER(uuid_d)
 ````

I don't want to go digging in `src/include/random.h` blindly.

**Comment by batrick** [comment]
Comparison before/after with the current patch of file descriptors open at start:

````
pdonnell@senta02 ~/ceph/build$ ls -l /proc/$(cat out/mds.a.pid)/fd 
total 0
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:50 0 -> /dev/null
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:50 1 -> /dev/null
lrwx------ 1 pdonnell pdonnell 64 Sep 12 22:50 10 -> socket:[237550793]
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:50 11 -> /dev/urandom
lrwx------ 1 pdonnell pdonnell 64 Sep 12 22:50 12 -> anon_inode:[eventpoll]
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:50 13 -> pipe:[237550798]
l-wx------ 1 pdonnell pdonnell 64 Sep 12 22:50 14 -> pipe:[237550798]
lrwx------ 1 pdonnell pdonnell 64 Sep 12 22:50 15 -> anon_inode:[eventpoll]
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:50 16 -> pipe:[237550799]
l-wx------ 1 pdonnell pdonnell 64 Sep 12 22:50 17 -> pipe:[237550799]
lrwx------ 1 pdonnell pdonnell 64 Sep 12 22:50 18 -> anon_inode:[eventpoll]
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:50 19 -> pipe:[237550800]
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:50 2 -> /dev/null
l-wx------ 1 pdonnell pdonnell 64 Sep 12 22:50 20 -> pipe:[237550800]
lrwx------ 1 pdonnell pdonnell 64 Sep 12 22:50 21 -> socket:[237516442]
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:50 22 -> pipe:[237550801]
l-wx------ 1 pdonnell pdonnell 64 Sep 12 22:50 23 -> pipe:[237550801]
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:50 24 -> pipe:[237550802]
l-wx------ 1 pdonnell pdonnell 64 Sep 12 22:50 25 -> pipe:[237550802]
lrwx------ 1 pdonnell pdonnell 64 Sep 12 22:50 26 -> socket:[237545851]
lrwx------ 1 pdonnell pdonnell 64 Sep 12 22:50 27 -> socket:[237541752]
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:50 28 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:50 29 -> pipe:[237541746]
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:50 3 -> /dev/urandom
l-wx------ 1 pdonnell pdonnell 64 Sep 12 22:50 30 -> pipe:[237541746]
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:50 31 -> pipe:[237541747]
l-wx------ 1 pdonnell pdonnell 64 Sep 12 22:50 32 -> pipe:[237541747]
lrwx------ 1 pdonnell pdonnell 64 Sep 12 22:50 33 -> socket:[237516445]
lrwx------ 1 pdonnell pdonnell 64 Sep 12 22:50 34 -> socket:[237523718]
lrwx------ 1 pdonnell pdonnell 64 Sep 12 22:50 35 -> socket:[237522295]
l-wx------ 1 pdonnell pdonnell 64 Sep 12 22:50 4 -> /home/pdonnell/ceph/build/out/mds.a.log
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:50 5 -> /dev/urandom
lrwx------ 1 pdonnell pdonnell 64 Sep 12 22:50 6 -> /home/pdonnell/ceph/build/out/mds.a.pid
lrwx------ 1 pdonnell pdonnell 64 Sep 12 22:50 7 -> socket:[237544772]
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:50 8 -> pipe:[237550792]
l-wx------ 1 pdonnell pdonnell 64 Sep 12 22:50 9 -> pipe:[237550792]
````

after a few dozen failovers with the current patch:

````
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 0 -> /dev/null
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 1 -> /dev/null
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 10 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 100 -> pipe:[237541046]
l-wx------ 1 pdonnell pdonnell 64 Sep 12 22:53 101 -> pipe:[237541046]
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 11 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 12 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 13 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 14 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 15 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 16 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 17 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 18 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 19 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 2 -> /dev/null
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 20 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 21 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 22 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 23 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 24 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 25 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 26 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 27 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 28 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 29 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 3 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 30 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 31 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 32 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 33 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 34 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 35 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 36 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 37 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 38 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 39 -> /dev/urandom
l-wx------ 1 pdonnell pdonnell 64 Sep 12 22:53 4 -> /home/pdonnell/ceph/build/out/mds.a.log
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 40 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 41 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 42 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 43 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 44 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 45 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 46 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 47 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 48 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 49 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 5 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 50 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 51 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 52 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 53 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 54 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 55 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 56 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 57 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 58 -> /dev/urandom
lrwx------ 1 pdonnell pdonnell 64 Sep 12 22:53 59 -> /home/pdonnell/ceph/build/out/mds.a.pid
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 6 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 60 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 61 -> /dev/urandom
lrwx------ 1 pdonnell pdonnell 64 Sep 12 22:53 62 -> socket:[237542099]
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 63 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 64 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 65 -> pipe:[237541033]
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 66 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 67 -> /dev/urandom
l-wx------ 1 pdonnell pdonnell 64 Sep 12 22:53 68 -> pipe:[237541033]
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 69 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 7 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 70 -> /dev/urandom
lrwx------ 1 pdonnell pdonnell 64 Sep 12 22:53 71 -> socket:[237541034]
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 72 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 73 -> /dev/urandom
lrwx------ 1 pdonnell pdonnell 64 Sep 12 22:53 74 -> anon_inode:[eventpoll]
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 75 -> pipe:[237541040]
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 76 -> /dev/urandom
l-wx------ 1 pdonnell pdonnell 64 Sep 12 22:53 77 -> pipe:[237541040]
lrwx------ 1 pdonnell pdonnell 64 Sep 12 22:53 78 -> anon_inode:[eventpoll]
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 79 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 8 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 80 -> pipe:[237541041]
l-wx------ 1 pdonnell pdonnell 64 Sep 12 22:53 81 -> pipe:[237541041]
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 82 -> /dev/urandom
lrwx------ 1 pdonnell pdonnell 64 Sep 12 22:53 83 -> anon_inode:[eventpoll]
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 84 -> pipe:[237541042]
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 85 -> /dev/urandom
l-wx------ 1 pdonnell pdonnell 64 Sep 12 22:53 86 -> pipe:[237541042]
lrwx------ 1 pdonnell pdonnell 64 Sep 12 22:53 87 -> socket:[237530647]
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 88 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 89 -> pipe:[237541043]
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 9 -> /dev/urandom
l-wx------ 1 pdonnell pdonnell 64 Sep 12 22:53 90 -> pipe:[237541043]
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 91 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 92 -> pipe:[237541044]
l-wx------ 1 pdonnell pdonnell 64 Sep 12 22:53 93 -> pipe:[237541044]
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 94 -> /dev/urandom
lrwx------ 1 pdonnell pdonnell 64 Sep 12 22:53 96 -> socket:[237544045]
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 97 -> /dev/urandom
lr-x------ 1 pdonnell pdonnell 64 Sep 12 22:53 98 -> pipe:[237541045]
l-wx------ 1 pdonnell pdonnell 64 Sep 12 22:53 99 -> pipe:[237541045]
````

**Comment by cbodley** [comment]
> Our use of `std::random_device`/`boost::random::random_device` also leaks a file descriptor to `/dev/urandom`. This is not easily fixed. @cbodley @adamemerson any suggestions?

i think this should be enough:
```patch
diff --git a/src/include/random.h b/src/include/random.h
index cbfee2d723..b3cb80c31d 100644
--- a/src/include/random.h
+++ b/src/include/random.h
@@ -91,7 +91,7 @@ void randomize_rng(const SeedT seed, MutexT& m, EngineT& e)
 template <typename MutexT, typename EngineT>
 void randomize_rng(MutexT& m, EngineT& e)
 {
-  thread_local std::random_device rd;
+  std::random_device rd;
  
   std::lock_guard<MutexT> lg(m);
   e.seed(rd());
@@ -107,7 +107,7 @@ void randomize_rng(const SeedT n)
 template <typename EngineT = std::default_random_engine>
 void randomize_rng()
 {
-  thread_local std::random_device rd;
+  std::random_device rd;
   detail::engine<EngineT>().seed(rd());
 }
 ```
there's no reason for those to be thread local, as they're only used to generate a seed on first use

**Comment by batrick** [comment]
> there's no reason for those to be thread local, as they're only used to generate a seed on first use

Well that was easy! Thanks for the suggestion. I think that fixed the last of the leaks. (There are still some races with urandom fds but they are small enough to not matter over a long period of time.)

**Comment by batrick** [comment]
@cbodley @tchaikov this looks good to go based on my testing. Please add your reviews when you have time.

**Comment by liewegas** [comment]
@wjwithagen

**Comment by ukernel** [comment]
how about closing file descriptors (/proc/self/fd/*) when ceph daemon starts

File: src/common/socket.c [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+#include <sys/socket.h>
+#include <sys/types.h>
+
+int socket_cloexec(int domain, int type, int protocol)
+{
+#ifdef SOCK_CLOEXEC
+  return socket(domain, type|SOCK_CLOEXEC, protocol);
+#else
+  int fd = socket(domain, type, protocol);
+  if (fd == -1)
```
Comment (tchaikov) [inline src/common/socket.c]: i'd suggest just return the return value of `socket()` or `fcntl()`.

File: src/common/socket.c [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+int socketpair_cloexec(int domain, int type, int protocol, int sv[2])
+{
+#ifdef SOCK_CLOEXEC
+  return socketpair(domain, type|SOCK_CLOEXEC, protocol, sv);
+#else
+  int rc = socketpair(domain, type, protocol, sv);
+  if (rc == -1)
+    return -errno;
```
Comment (tchaikov) [inline src/common/socket.c]: ditto.

File: src/common/socket.c [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+#include "include/compat.h"
+
+#include <errno.h>
+#include <fcntl.h>
+
+#include <sys/socket.h>
+#include <sys/types.h>
+
+int socket_cloexec(int domain, int type, int protocol)
```
Comment (tchaikov) [inline src/common/socket.c]: could also just define these functions in `compat.cc`. especially `accept4()`. it's easy to have a userspace implementation of it if it's not around.

**Comment by tchaikov** [comment]
> how about closing file descriptors (/proc/self/fd/*) when ceph daemon starts

`/proc` fs is not portable. probably we can avoid using it if possible.

File: src/common/socket.c (hunk 1 above)
Comment (batrick) [reply src/common/socket.c]: I'm following the style of `pipe_cloexec` for consistency.

Comment (batrick) [reply src/common/socket.c]: I think I'll change the API here and in `pipe_cloexec` though as it is confusing and I htink the return code checking is wrong in this PR.

**Comment by batrick** [comment]
retest this please

**Comment by batrick** [comment]
@tchaikov please have another look

**Comment by batrick** [comment]
Rebased to resolved conflicts.

**Comment by batrick** [comment]
Rebased to resolve conflicts.

**Comment by tchaikov** [comment]
@batrick needs rebase.

**Comment by batrick** [comment]
@tchaikov  done

**Comment by tchaikov** [comment]
retest this please.

**Comment by batrick** [comment]
tests pass after rebase (probably coincidence)

**Comment by batrick** [comment]
http://pulpito.ceph.com/?branch=wip-pdonnell-testing-20180924.230702

File: src/include/uuid.h [hunk 4]
```diff
@@ -26,8 +27,8 @@ struct uuid_d {
   }
 
   void generate_random() {
-    boost::random::random_device rng("/dev/urandom");
-    boost::uuids::basic_random_generator<boost::random::random_device> gen(&rng);
+    std::random_device rng;
+    boost::uuids::basic_random_generator gen(rng);
```
Comment (thinkercui) [inline src/include/uuid.h]: mybe basic_random_generator need a template argument. it should be `boost::uuids::basic_random_generator<std::random_device>`

Comment (tchaikov) [reply src/include/uuid.h]: in C++17, it can be deduced. BTW, the change has been merged. you might want to file a ticket a PR if you believe it is a bug.

Comment (thinkercui) [reply src/include/uuid.h]: @tchaikov  thanks a lot. the newer the better. we will just fix it at our own branch.

**Pull request merged**